### ISSUE TITLE
In 1.3, kubelet cloud-provider setting is not respected when --cloud-provider=""

### ISSUE THREAD

**Issue by emaildanwilson** [kubernetes/kubernetes#28231]
It appears that the kubelet will detect and use the cloud provider as detected by cadvisor thus overriding the cloud provider passed in at runtime: https://github.com/kubernetes/kubernetes/blob/release-1.3/pkg/kubelet/kubelet.go#L1117. This will break our private cloud configuration where we do not use our cloud provider's load balancing and instead use our own load balancing tier. In my opinion, the cloud provider specified at kubelet runtime should be respected for the node.

Am I reading this correctly?
Any thoughts on how to handle this properly?

Thank you!

**Comment by yujuhong** [comment]
If you specify any cloud provider through flags, it will be used by kubelet (see https://github.com/kubernetes/kubernetes/blob/release-1.3/pkg/kubelet/kubelet.go#L1068). It's only when you *don't* specify any cloud provider that kubelet will detect the cloud provider from the machine info, as described by the comment.
```go
		// If no cloud provider is defined - use the one detected by cadvisor
```

**Comment by emaildanwilson** [comment]
that's exactly my situation, I want to run in our private cloud environment without any of the cloud integration.

**Comment by yujuhong** [comment]
> that's exactly my situation, I want to run in our private cloud environment without any of the cloud integration. If I specify cloud-provider="" at runtime it will get overridden.

Ah I see. I guess specifying an empty string is the same as not specifying any cloud provider. 

This auto-detection feature was introduced by #21373
/cc the author and reviewer to see if this is an intended behavior: @enoodle @vishh

**Comment by vishh** [comment]
@emaildanwilson Legitimate concern. I'm posting a fix. Do you consider this to be a regression or are you just getting started with k8s.

**Comment by emaildanwilson** [comment]
@vishh Thank you. We're running on 1.2.4 and won't be running on 1.3 for a few months yet anyways and was just testing to look for future issues when I found this.

**Comment by yujuhong** [comment]
@vishh I assigned this issue to you since [you are working on a fix](https://github.com/kubernetes/kubernetes/issues/28231#issuecomment-229476178), and I can't assign this to @enoodle anyway :-)

**Comment by dchen1107** [comment]
@emaildanwilson Thanks for testing our 1.3 release candidate, and reporting this regression to us. We are including the fix to 1.3.
